### PR TITLE
Add Percentile support

### LINKED ISSUE #72: Feature request: quantile/percentile based statistics
First of all, congratulations: I'm seriously considering making ddosify part of my daily stack.
Here's what I see missing. 
When measuring latencies it is common practice to measure a certain percentile. It is also part of the SLA terminology, which this tool could help to assess.
However, I didn't find a way to get a latency statistical distribution.
Are you perhaps planning to add it?
Cheers

### PR DESCRIPTION
**Pull request by zhiburt**
Hi there

I doubt I've done the right thing...
But I tried.

image

close #72

### CODE REVIEW

**Comment by zhiburt** [comment]
I guess it should consider all scenarious?

image

**Review by kursataktas — COMMENTED** [review]
Thank you @zhiburt for this work. I've added my comments.

1.  I think we don't need to display percentiles on realtime data. 
2.  I guess p99, p95, p90, and p80 would be enough for now.
3. Let's add a new flag to display/calculate percentiles as optional. Since percentile calculation would be time-consuming.

Example (CLI):
`ddosify -t target.com --output-percentile`

Example (Config JSON):
```
{
    "request_count": 10,
    "load_type": "linear",
    "duration": 5,
    "output": "stdout-json",
    "output_percentile": true,
    ...
}
```
You may need to change the [Init()](https://github.com/ddosify/ddosify/blob/master/core/report/base.go#L35) contract of the Report service to pass the new flag to report service implementations.



**Example outputs on different scenarios;**

- 1 Step stdout
image

- Multi Step stdout
image

- 1 Step stdout-json

```
{
    "success_perc": 100,
    "fail_perc": 0,
    "success_count": 10,
    "fail_count": 0,
    "avg_duration": 0.225,
    "percentiles": {
          "p99": 0.123,
          "p95": 0.321,
          "p90": 0.33,
          "p80": 0.51
     },
    "steps": {
      "1": {
        "name": "D-1",
        "status_code_dist": {
          "200": 10
        },
        "error_dist": {
          
        },
        "durations": {
          "connection": 0.009,
          "dns": 0.001,
          "request_write": 0,
          "response_read": 0,
          "server_processing": 0.095,
          "tls": 0.01,
          "total": 0.116
        },
        "percentiles": {  // all percentiles are calculated on durations.total value.
          "p99": 0.123,
          "p95": 0.321,
          "p90": 0.33,
          "p80": 0.51
        },
        "success_count": 10,
        "fail_count": 0,
        "success_perc": 100,
        "fail_perc": 0
      }
    }
  }
```

- Multi Step stdout-json
```
{
  "success_perc": 100,
  "fail_perc": 0,
  "success_count": 10,
  "fail_count": 0,
  "avg_duration": 0.225,
  "total_duration_percentiles": {
        "p99": 0.123,
        "p95": 0.321,
        "p90": 0.33,
        "p80": 0.51
   },
  "steps": {
    "1": {
      "name": "D-1",
      "status_code_dist": {
        "200": 10
      },
      "error_dist": {
        
      },
      "durations": {
        "connection": 0.009,
        "dns": 0.001,
        "request_write": 0,
        "response_read": 0,
        "server_processing": 0.095,
        "tls": 0.01,
        "total": 0.116
      },
      "total_duration_percentiles": {  // all percentiles are calculated on durations.total value.
        "p99": 0.123,
        "p95": 0.321,
        "p90": 0.33,
        "p80": 0.51
      },
      "success_count": 10,
      "fail_count": 0,
      "success_perc": 100,
      "fail_perc": 0
    },
    "2": {
      "name": "D-2",
      "status_code_dist": {
        "200": 10
      },
      "error_dist": {
        
      },
      "durations": {
        "connection": 0.008,
        "dns": 0.001,
        "request_write": 0,
        "response_read": 0,
        "server_processing": 0.09,
        "tls": 0.009,
        "total": 0.109
      },
      "total_duration_percentiles": {  // all percentiles are calculated on durations.total value.
        "p99": 0.123,
        "p95": 0.321,
        "p90": 0.33,
        "p80": 0.51
      },
      "success_count": 10,
      "fail_count": 0,
      "success_perc": 100,
      "fail_perc": 0
    }
  }
}
```

I wonder what you think about these?

**Comment by zhiburt** [comment]
Everything must be addressed.

Though I haven't tested json output.

image

File: core/report/stdoutJson.go [hunk 1]
```diff
@@ -54,12 +57,57 @@ func (s *stdoutJson) Start(input chan *types.Response) {
...
+			StatusCodeDist: item.StatusCodeDist,
+			ErrorDist:      item.ErrorDist,
+			Durations:      item.Durations,
+			SuccessCount:   item.SuccessCount,
+			FailedCount:    item.FailedCount,
+		}
+
+		if s.reportPercentiles {
+			jsonResult.ItemReports[key].Percentiles = map[string]float32{
+				"p99": item.DurationPercentile(99),
```
Comment (kursataktas) [inline core/report/stdoutJson.go:97]: Milliseconds would be enough. 

Current stdout-json example:
```
...
"percentiles": {
        "p80": 0.12210735, --> "0.122" would be enough
        "p90": 0.14628321,
        "p95": 0.1905153,
        "p99": 0.1905153
      },
...
```

File: core/report/stdoutJson.go [hunk 2]
```diff
@@ -54,12 +57,57 @@ func (s *stdoutJson) Start(input chan *types.Response) {
...
+		jsonResult.ItemReports[key] = &jsonScenarioItemReport{
+			Name:           item.Name,
+			StatusCodeDist: item.StatusCodeDist,
+			ErrorDist:      item.ErrorDist,
+			Durations:      item.Durations,
+			SuccessCount:   item.SuccessCount,
+			FailedCount:    item.FailedCount,
+		}
+
+		if s.reportPercentiles {
```
Comment (kursataktas) [inline core/report/stdoutJson.go:95]: If possible, let's not show the percentile key on if it is not enabled. 

Current state:
```
...
 "durations": {
        "connection": 0.005,
        "dns": 0.009,
        "request_write": 0,
        "response_read": 0.382,
        "server_processing": 0.253,
        "tls": 0.012,
        "total": 0.661
      },
      "percentiles": null, ---> Let's remove this if we can
      "success_count": 10,
      "fail_count": 0
    },
...
```

**Review by kursataktas — COMMENTED** [review]
Thank you for this new design. It looks great 💯 

- I put my words for the related sections
- There are some failed tests, we have to take a look at them
- We should add more tests to cover new changes
- We may want to update the [example_config](https://github.com/ddosify/ddosify/blob/master/config_examples/config.json) to show the usage.

**Review by pelohsoe — APPROVED** [review]

**Pull request closed without merging**